src/main/java/net/neoforged/neoforge/common/damagesource/DamageContainer.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.damagesource;
+
+import java.util.AbstractMap;
+import java.util.Arrays;
+import java.util.EnumMap;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+import net.minecraft.world.damagesource.DamageSource;
+import net.minecraft.world.entity.LivingEntity;
+import net.neoforged.neoforge.event.entity.EntityInvulnerabilityCheckEvent;
+import net.neoforged.neoforge.event.entity.living.LivingDamageEvent;
+import net.neoforged.neoforge.event.entity.living.LivingIncomingDamageEvent;
+import net.neoforged.neoforge.event.entity.living.LivingShieldBlockEvent;
+
+/**
+ * DamageContainer encapsulates aspects of the entity damage sequence so that
+ * relevant context related to damage dealt is accessible throughout the entire
+ * sequence.
+ * <h4>Note: certain values will be defaults until the stage in the sequence when they are set.</h4>
+ * <H3>Damage Sequence and uses</H3><ol>
+ * <li>Entity is hurt by a damage source</li>
+ * <li>{@link EntityInvulnerabilityCheckEvent EntityInvulnerablityCheckEvent}
+ * fires and determines if the sequence can commence</li>
+ * <li>{@link LivingIncomingDamageEvent EntityPreDamageEvent} fires
+ * and gives access to this. Modifiers should be added here.</li>
+ * <li>{@link LivingShieldBlockEvent} fires</li>
+ * <li>armor, enchantments, mob effect, and absorption modifiers are applied to the damage</li>
+ * <li>{@link LivingDamageEvent.Pre IncomingDamageEvent} fires and
+ * provides final values for the preceding modifiers and the last chance to negate the damage but will not
+ * undo the preceding effects</li>
+ * <li>{@link LivingDamageEvent.Post DamageTakenEvent} fires and provides
+ * an immutable perspective of what the entire sequence ended with. </li>
+ * </ol>
+ *
+ *
+ */
+public interface DamageContainer {
```

Comment:
Since you're dealing with both mutable and immutable containers, I feel like you should have a super `DamageContainer` and then a `MutableDamageContainer` to indicate the differenece more clearly.